packages/completer/src/connectorproxy.ts

```diff
@@ -0,0 +1,30 @@
+import { CompletionHandler } from './handler';
+import { IConnectorProxy } from './tokens';
+export class ConnectorProxy implements IConnectorProxy {
+  constructor(connectorMap: ConnectorProxy.IConnectorMap) {
+    this._connectorMap = connectorMap;
+  }
+
+  public async fetch(
+    request: CompletionHandler.IRequest
+  ): Promise<Array<{ [id: string]: CompletionHandler.ICompletionItemsReply }>> {
+    let promises: Promise<{
+      [id: string]: CompletionHandler.ICompletionItemsReply;
+    }>[] = [];
+    for (const [id, connector] of this._connectorMap.entries()) {
+      let promise = connector.fetch(request).then(reply => ({ [id]: reply }));
+      promises.push(promise.catch(p => p));
+    }
+    const combinedPromise = Promise.all(promises);
```

Comment:
This would be alleviated by https://github.com/jupyterlab/jupyterlab/pull/11831 :)

Reply:
Hi @krassowski, thanks for your review!

> * have a timeout for kernel completions (because kernel can take a long time to reply): this can be simply implemented by wrapping each promise with `promise = Promise.race(promise, resolveOnTimeout(settings.connectorTimeout))`
> * do not ask kernel for completions in the first place when the kernel is busy; similarly, the LSP connection may be temporarily disconnected due to network issues so this provider would be registered but temporarily unavailable. This requires changes to the interface; we would need to either:
>   
>   * reject in the base provider when kernel is busy (which means we would want to split the base provider into `contextProvider` and  `kernelProvider`), or
>   * have a method on provider interface to ask the provider "hey, provider are you available to provide completions now?", say `provider.isAvailable`
> 

these are good ideas, as I asked above, I'm not entirely sure about the provider interface `ICompletionProvider`, any suggestions about the attributes you need for a provider are very helpful.